File: backend-node/models/Product.js
const mongoose = require('mongoose');

const ProductSchema = new mongoose.Schema({
  name: {
    type: String,
    required: [true, 'Product name is required'],
    trim: true,
    maxlength: [100, 'Product name cannot be more than 100 characters']
  },
  description: {
    type: String,
    required: [true, 'Product description is required'],
    maxlength: [1000, 'Description cannot be more than 1000 characters']
  },
  price: {
    type: Number,
    required: [true, 'Product price is required'],
    min: [0, 'Price cannot be negative']
  },
  originalPrice: {
    type: Number,
    required: [true, 'Original price is required'],
    min: [0, 'Original price cannot be negative']
  },
  category: {
    type: String,
    required: [true, 'Product category is required'],
    enum: [
      'Women Ethnic',
      'Women Western', 
      'Men',
      'Bags & Footwear',
      'Kids',
      'Home & Kitchen',
      'Beauty & Health',
      'Electronics'
    ]
  },
  images: [{
    type: String,
    required: true
  }],
  seller: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  rating: {
    type: Number,
    default: 0,
    min: [0, 'Rating cannot be less than 0'],
    max: [5, 'Rating cannot be more than 5']
  },
  reviews: {
    type: Number,
    default: 0,
    min: [0, 'Reviews count cannot be negative']
  },
  inventory: {
    type: Number,
    required: [true, 'Inventory count is required'],
    min: [0, 'Inventory cannot be negative'],
    default: 100
  },
  sizes: [{
    type: String,
    enum: ['XS', 'S', 'M', 'L', 'XL', 'XXL', 'Free Size']
  }],
  colors: [{
    type: String
  }],
  deliveryTime: {
    type: String,
    default: '7-10 days'
  },
  freeDelivery: {
    type: Boolean,
    default: true
  },
  isActive: {
    type: Boolean,
    default: true
  },
  tags: [{
    type: String,
    trim: true
  }]
}, {
  timestamps: true
});

// Virtual for discount percentage
ProductSchema.virtual('discount').get(function() {
  if (this.originalPrice > this.price) {
    return Math.round(((this.originalPrice - this.price) / this.originalPrice) * 100);
  }
  return 0;
});

// Ensure virtual fields are serialized
ProductSchema.set('toJSON', { virtuals: true });

// Index for search functionality
ProductSchema.index({ 
  name: 'text', 
  description: 'text', 
  category: 'text',
  tags: 'text'
});

// Index for filtering
ProductSchema.index({ category: 1, price: 1, rating: 1 });
ProductSchema.index({ seller: 1 });

module.exports = mongoose.model('Product', ProductSchema);

File: backend/models/product.py
from pydantic import BaseModel, Field
from typing import List, Optional
from datetime import datetime
from bson import ObjectId
from .user import PyObjectId

class ProductCreate(BaseModel):
    name: str
    description: str
    price: float
    originalPrice: float
    category: str
    images: List[str] = []
    sizes: List[str] = []
    colors: List[str] = []
    deliveryTime: str = "7-10 days"
    freeDelivery: bool = True
    inventory: int = 100

class Product(BaseModel):
    id: PyObjectId = Field(default_factory=PyObjectId, alias="_id")
    name: str
    description: str
    price: float
    originalPrice: float
    category: str
    images: List[str] = []
    seller: PyObjectId
    rating: float = 0.0
    reviews: int = 0
    inventory: int = 100
    sizes: List[str] = []
    colors: List[str] = []
    deliveryTime: str = "7-10 days"
    freeDelivery: bool = True
    isActive: bool = True
    createdAt: datetime = Field(default_factory=datetime.utcnow)
    updatedAt: datetime = Field(default_factory=datetime.utcnow)
    
    class Config:
        allow_population_by_field_name = True
        arbitrary_types_allowed = True
        json_encoders = {ObjectId: str}

class ProductResponse(BaseModel):
    id: str
    name: str
    description: str
    price: float
    originalPrice: float
    category: str
    images: List[str]
    seller: str
    rating: float
    reviews: int
    inventory: int
    sizes: List[str]
    colors: List[str]
    deliveryTime: str
    freeDelivery: bool
    discount: int = 0
    
    def __init__(self, **data):
        super().__init__(**data)
        if self.originalPrice > self.price:
            self.discount = int(((self.originalPrice - self.price) / self.originalPrice) * 100)

class ProductUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    price: Optional[float] = None
    originalPrice: Optional[float] = None
    category: Optional[str] = None
    images: Optional[List[str]] = None
    inventory: Optional[int] = None
    sizes: Optional[List[str]] = None
    colors: Optional[List[str]] = None
    deliveryTime: Optional[str] = None
    freeDelivery: Optional[bool] = None

class Category(BaseModel):
    id: int
    name: str
    icon: str
    image: str